ml-algorithms/src/main/java/org/opensearch/ml/engine/tools/QueryPlanningTool.java

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.ml.engine.tools;
+
+import java.util.List;
+import java.util.Map;
+
+import org.apache.commons.text.StringSubstitutor;
+import org.opensearch.core.action.ActionListener;
+import org.opensearch.ml.common.spi.tools.ToolAnnotation;
+import org.opensearch.ml.common.spi.tools.WithModelTool;
+import org.opensearch.ml.repackage.com.google.common.annotations.VisibleForTesting;
+import org.opensearch.transport.client.Client;
+
+import lombok.Getter;
+import lombok.Setter;
+
+/**
+ * This tool supports different types of query planning,
+ * llmGenerated, systemSearchTemplates or userSearchTemplates.
+ * //TODO only support llmGenerated for now.
+ * //TODO to add in systemSearchTemplates or userSearchTemplates when searchTemplatesTool is implemented.
+ */
+
+@ToolAnnotation(QueryPlanningTool.TYPE)
+public class QueryPlanningTool implements WithModelTool {
+    public static final String TYPE = "QueryPlanningTool";
+    public static final String MODEL_ID_FIELD = "model_id";
+    private final MLModelTool queryGenerationTool;
+    public static final String PROMPT_FIELD = "prompt";
```

Comment:
Should we also support `system_prompt` ?

Reply:
right, we should merge this first, @rithin-pullela-aws 's PR to add in prompt and feature flag is depending on this one.